test/lit/passes/cast-refining.wast

```diff
@@ -0,0 +1,1039 @@
+        (local.get $x)
+      )
+    )
+    (drop
+      (ref.cast null $C1
+        (local.get $x)
+      )
+    )
+    (drop
+      (ref.cast null $C2
+        (local.get $x)
+      )
+    )
+  )
+
+  ;; YESTNH:      (func $ref.test (type $anyref_=>_none) (param $x anyref)
+  ;; YESTNH-NEXT:  (drop
+  ;; YESTNH-NEXT:   (block (result i32)
+  ;; YESTNH-NEXT:    (drop
+  ;; YESTNH-NEXT:     (ref.test $A
+  ;; YESTNH-NEXT:      (local.get $x)
+  ;; YESTNH-NEXT:     )
+  ;; YESTNH-NEXT:    )
+  ;; YESTNH-NEXT:    (i32.const 0)
+  ;; YESTNH-NEXT:   )
+  ;; YESTNH-NEXT:  )
+  ;; YESTNH-NEXT:  (drop
+  ;; YESTNH-NEXT:   (ref.is_null
+  ;; YESTNH-NEXT:    (local.get $x)
+  ;; YESTNH-NEXT:   )
+  ;; YESTNH-NEXT:  )
+  ;; YESTNH-NEXT: )
+  ;; NO_TNH:      (func $ref.test (type $anyref_=>_none) (param $x anyref)
+  ;; NO_TNH-NEXT:  (drop
+  ;; NO_TNH-NEXT:   (block (result i32)
+  ;; NO_TNH-NEXT:    (drop
+  ;; NO_TNH-NEXT:     (ref.test $A
+  ;; NO_TNH-NEXT:      (local.get $x)
+  ;; NO_TNH-NEXT:     )
+  ;; NO_TNH-NEXT:    )
+  ;; NO_TNH-NEXT:    (i32.const 0)
+  ;; NO_TNH-NEXT:   )
+  ;; NO_TNH-NEXT:  )
+  ;; NO_TNH-NEXT:  (drop
+  ;; NO_TNH-NEXT:   (ref.is_null
+  ;; NO_TNH-NEXT:    (local.get $x)
+  ;; NO_TNH-NEXT:   )
+  ;; NO_TNH-NEXT:  )
+  ;; NO_TNH-NEXT: )
+  (func $ref.test (param $x anyref)
+    ;; This will return 0.
+    (drop
+      (ref.test $A
+        (local.get $x)
+      )
+    )
+    ;; This can test for a null.
+    (drop
+      (ref.test null $A
+        (local.get $x)
+      )
+    )
+  )
+
+  ;; YESTNH:      (func $br_on (type $anyref_=>_none) (param $x anyref)
+  ;; YESTNH-NEXT:  (drop
+  ;; YESTNH-NEXT:   (block $block (result (ref $B))
+  ;; YESTNH-NEXT:    (drop
+  ;; YESTNH-NEXT:     (br_on_cast $block $B
+  ;; YESTNH-NEXT:      (local.get $x)
+  ;; YESTNH-NEXT:     )
+  ;; YESTNH-NEXT:    )
+  ;; YESTNH-NEXT:    (unreachable)
+  ;; YESTNH-NEXT:   )
+  ;; YESTNH-NEXT:  )
+  ;; YESTNH-NEXT:  (drop
+  ;; YESTNH-NEXT:   (block $block0 (result (ref any))
+  ;; YESTNH-NEXT:    (br_on_non_null $block0
+  ;; YESTNH-NEXT:     (local.get $x)
+  ;; YESTNH-NEXT:    )
+  ;; YESTNH-NEXT:    (unreachable)
+  ;; YESTNH-NEXT:   )
+  ;; YESTNH-NEXT:  )
+  ;; YESTNH-NEXT: )
+  ;; NO_TNH:      (func $br_on (type $anyref_=>_none) (param $x anyref)
+  ;; NO_TNH-NEXT:  (drop
+  ;; NO_TNH-NEXT:   (block $block (result (ref $A))
+  ;; NO_TNH-NEXT:    (drop
+  ;; NO_TNH-NEXT:     (br_on_cast $block $A
+  ;; NO_TNH-NEXT:      (local.get $x)
+  ;; NO_TNH-NEXT:     )
+  ;; NO_TNH-NEXT:    )
+  ;; NO_TNH-NEXT:    (unreachable)
+  ;; NO_TNH-NEXT:   )
+  ;; NO_TNH-NEXT:  )
+  ;; NO_TNH-NEXT:  (drop
+  ;; NO_TNH-NEXT:   (block $block0 (result (ref any))
+  ;; NO_TNH-NEXT:    (br_on_non_null $block0
+  ;; NO_TNH-NEXT:     (local.get $x)
+  ;; NO_TNH-NEXT:    )
+  ;; NO_TNH-NEXT:    (unreachable)
+  ;; NO_TNH-NEXT:   )
+  ;; NO_TNH-NEXT:  )
+  ;; NO_TNH-NEXT: )
+  (func $br_on (param $x anyref)
+    ;; We don't optimize here yet. TODO
```

Comment:
The $A is optimized to $B, so it would be good to clarify what optimization is missing.

Reply:
Fixed. I changed the test to focus on the optimization we don't do, and clarify that.